### PR TITLE
feat: support dspark for qwen3 on NPU.

### PR DESCRIPTION
**Pull request by yinjiawei01**
## Description

This PR introduces DSpark speculative decoding for the NPU/ATB backend.

DSpark consumes intermediate hidden states captured from the target model, performs one non-causal block-diffusion forward pass to generate base logits for the complete block, and then samples each position sequentially through a low-rank Markov head:

```text
markov_bias(t[i-1]) = embedding(t[i-1], markov_w1) @ markov_w2^T
draft_logits[i] = base_logits[i] + markov_bias(t[i-1])
```

The target model validates the complete proposal block with the existing shared rejection sampler, accepting the longest matching prefix plus one bonus token.

### Supported Models

- Targets: dense Qwen3 and Qwen3-MoE models with standard attention, `head_dim=128`, and a four-dimensional dense KV cache.
- Draft: always a small dense Qwen3 model registered as `DSparkDraftModel`; it never executes experts.
- Performance validation currently covers a dense Qwen3-8B target. A Qwen3-MoE benchmark has not yet been added.
- Hybrid or linear-attention targets, reduced-vocabulary drafts, and `cp_size > 1` are not supported.

### Implementation

- Adds `DSparkWorkerImpl` as a subclass of `DFlashWorkerImpl`, reusing target prefill, auxiliary-hidden capture, context-KV handling, validation, rejection sampling, and cache writeback.
- Uses an N-wide DSpark query in which the anchor predicts the first token and the remaining positions contain mask tokens. DFlash continues to use an N+1-wide query.
- Calculates the KV length as `original_kv_len + query_width - 1`, preventing non-causal attention from reading an uninitialized cache slot.
- Adds `DSparkDraftWorkerImpl` for logits-only draft forward, sequential Markov sampling, and TP token synchronization, keeping the DSpark lifecycle out of the generic `LLMWorkerImpl`.
- Adds the model-side `DSparkMarkovHead`, loading `markov_head.markov_w1.weight`, `markov_head.markov_w2.weight`, and the top-level `markov_rank`, while owning only low-rank bias computation.
- Registers `DSparkQwen3Model`, `DSparkQwen3ForCausalLM`, and `DSparkDraftModel`, and supports `--speculative_algorithm=DSpark`.
- Supports both top-level and nested `dflash_config` values for `target_layer_ids` and `mask_token_id`.

### Sampling and Tensor Parallelism

- Supports greedy, random, and mixed greedy/random sampling within the same batch.
- Greedy proposal probabilities are set to one. Random or mixed sampling returns the proposal probability of the sampled token.
- Markov weights are replicated on every TP rank, while the draft backbone LM head performs its normal TP gather once for the complete block.
- Greedy sampling requires no additional token broadcast. Random or mixed sampling broadcasts the sampled token after each position to keep the Markov chains synchronized.
- Single-node TP=2 has been validated with one xLLM process per NPU. TP=2 is included only as a compatibility check and is not part of the performance comparison.

### Requirements and Limitations

- `--enable_chunked_prefill=true` is required because the NPU Qwen3 path generates the non-causal block mask required by DSpark only on the chunked-prefill path. The server fails fast when the option is disabled.
- `cp_size > 1` is not supported because the context-parallel path cannot yet provide the correct target auxiliary hidden states.
- Target and draft vocabulary sizes must match. Reduced-vocabulary token remapping is not implemented.
- Draft sliding-window settings are currently ignored by the NPU FIA path, matching the existing DFlash behavior.
- Non-causal draft queries use eager execution. Fixed-shape verify graph capture is left as follow-up work.

### Validation

The benchmark uses Qwen3-8B, GSM8K, 128 prompts, greedy sampling, and `max_new_tokens=2048`. The client uses EvalScope with an OpenAI-compatible streaming API. The server enables continuous batching, prefix cache, chunked prefill, and schedule overlap. DSpark uses `num_speculative_tokens=15` and KV cache `block_size=128`.

All nine TP=1 test groups completed with 128 successful requests and zero failures.

| Concurrency | Baseline TPOT (ms) | Baseline E2E (s) | DFlash TPOT (ms) | DFlash E2E (s) | DFlash Accept Length | DSpark TPOT (ms) | DSpark E2E (s) | DSpark Accept Length | Baseline → DSpark | DFlash → DSpark |
|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|---:|
| 1 | 13.49 | 4.3258 | 3.32 | 1.0817 | 6.1192 | 2.97 | 0.9830 | 7.1643 | 4.40x | 1.10x |
| 8 | 15.13 | 4.7288 | 5.12 | 1.7134 | 6.1160 | 4.54 | 1.5305 | 7.2179 | 3.09x | 1.12x |
| 32 | 17.99 | 5.7469 | 11.21 | 3.7581 | 6.1511 | 9.81 | 3.3320 | 7.1799 | 1.72x | 1.13x |

- The DSpark accept length remains stable between 7.16 and 7.22, compared with 6.12 to 6.15 for DFlash.
- DSpark improves E2E latency by approximately 1.10x to 1.13x over DFlash and 1.72x to 4.40x over the target-only baseline.
- At higher concurrency, target validation, scheduling, and batching overhead reduce the relative speculative-decoding benefit.

## Related Issues

N/A.

## Change Type

- [ ] Bug fix
- [x] New feature
- [ ] Performance improvement
- [ ] Refactor
- [ ] Documentation
- [ ] Test
- [ ] Build or CI

## Pull Request Checklist

Thank you for contributing to xLLM. Before requesting review, please make sure the following items are complete.

### PR Title and Commit Messages

- [x] The PR title and each commit message follow the xLLM commit format: `<type>: <subject>`.

> Allowed types: feat, bugfix, docs, test, refactor, chore, style, revert, perf, model, build, release.
> The subject should use clear English, start with a verb, include at least 4 words, and end with `.`.

### Pre-commit Checks

- [x] I have installed `pre-commit` by running `pip install pre-commit` or an equivalent command.
- [x] I have installed the hooks with `pre-commit install`.
- [ ] I have run `pre-commit run --all-files` and fixed any reported issues.

> If you are unsure how to set up `pre-commit`, see [the pre-commit documentation](https://pre-commit.com/).

### Self Review

- [ ] I have self-reviewed the code according to `.agents/skills/code-review/references/custom-code-style.md`, especially code written or assisted by AI.
- [x] I have rebased this PR onto the latest `main` branch.

### Build and Test Coverage

- [ ] Tests have been added or updated as needed.
- [ ] CUDA: `python setup.py build test` has passed on a CUDA machine.
- [x] NPU: `python setup.py build test` has passed on an NPU machine.
- [ ] MLU: `python setup.py build test` has passed on an MLU machine.

### CODE REVIEW

**Review by weizhehuang0827 — DISMISSED** [review]

File: xllm/core/runtime/llm_worker_impl.cpp [hunk 1]
```diff
@@ -175,6 +175,61 @@ std::optional<ForwardOutput> LLMWorkerImpl::execute_no_sync_on_stream(
...
+  BlockDraftSampler* block_draft_sampler = model_->block_draft_sampler();
+  CHECK(block_draft_sampler != nullptr)
+      << "The draft model does not provide a block draft sampler.";
+
+  BlockDraftSampleOutput sample_output;
+  {
+    c10::StreamGuard stream_guard = compute_stream.set_stream_guard();
+    sample_output = block_draft_sampler->sample(
+        base_logits, anchor_token_ids, sampling_params);
+  }
```
[the remark this answers is not in the document]
Comment (yinjiawei01) [reply xllm/core/runtime/llm_worker_impl.cpp]: Fixed.

**Review by ustcfy — DISMISSED** [review]

**Review by weizhehuang0827 — APPROVED** [review]

**Review by ustcfy — APPROVED** [review]
